dev-docs/design/how-lit-html-works.md

```diff
@@ -0,0 +1,425 @@
+
+```js
+{
+  _$litType$: 1,
+  strings: [
+    '<span class="',
+    '">',
+    '</span><button @click=',
+    '>Increment</button>'
+  ],
+  values: [
+    "odd",
+    1,
+    () => render(counterUi(2), container)
+  ]
+}
+```
+
+Instead of redoing the work in the **Prepare** phase, the Lit `Template` is fetched from the cache using the template's `strings` Template Strings Array.
+
+> **Note**
+> The reference to the Template Strings Array is the cache key, not the value of the key. This is an O(1) lookup and not changed by the length of the static contents.
+
+The container's `ChildPart` stores the last committed value - which in this case is a `TemplateInstance` referencing the `Template` returned from the cache. Because we're rendering the same `Template` to this `ChildPart` the **Create** phase is skipped.
+
+All that is left is to iterate over the previously created `Part`s and call `_$setValue` with the new values. For demonstrative purposes, inlined this looks like:
+
+```js
+AttributePart._$setValue('odd');
+ChildPart._$setValue(1);
+EventPart._$setValue(() => render(counterUi(2), container));
+```
+
+This results in the DOM updating with a green underlined number one, and the event listener is swapped with the new listener.
+
+# Future work
+
+lit-html is fast, but what if it could be faster!?
+
+- [Declarative syntax for creating DOM Parts](https://github.com/WICG/webcomponents/issues/990). This could remove the requirement for lit-html to walk the `<template>` element's tree and manually instantiate Parts. This reduces the tree walk done in the **Create** phase.
+- Precompiling lit-html `html` tag functions (tracked in https://github.com/lit/lit/issues/189). This would allow opting into doing the **Prepare** phase in a compile step done by build tooling. Runtime would no longer have a **Prepare** phase.
+
+# Appendix
+
+## Parts
+
+A `Part` is a lit-html concept and represents the location of an expression in the `html` tagged template literal:
+
+| Part                   | Description                                                        | Authored                               |
+| ---------------------- | ------------------------------------------------------------------ | -------------------------------------- |
+| `ChildPart`            | Expressions in HTML child position                                 | `` html`<div>${...}</div>`  ``         |
+| `AttributePart`        | Expressions in HTML attribute value position                       | `` html`<input id="${...}">`  ``       |
+| `BooleanAttributePart` | Expressions in a boolean attribute value (name prefixed with `?`)  | `` html`<input ?checked="${...}">`  `` |
+| `EventPart`            | Expressions in an event listener position (name prefixed with `@`) | `` html`<input @click=${...}>`  ``     |
+| `PropertyPart`         | Expressions in property value position (name prefixed with `.`)    | `` html`<input .value=${...}>`  ``     |
```

Comment:
could we swap the order here so it's attribute, boolean, attribute, property, then event? To me that progression makes more logical sense as attribute is base, boolean attribute is a special case of that, property deviating a bit further, and event being quite different. It is also the order used here: https://lit.dev/docs/templates/expressions/

Also wondering if we might change the `<input>` to a `<button>` for the event? It seems like the more canonical element to be clicked. No big deal if we prefer the consistency of the same element.

```suggestion
| `PropertyPart`         | Expressions in property value position (name prefixed with `.`)    | `` html`<input .value=${...}>`  ``     |
| `EventPart`            | Expressions in an event listener position (name prefixed with `@`) | `` html`<button @click=${...}>`  ``     |
```

Reply:
I like it! (Slightly modified suggestion to add closing `</button>` tag).